### ISSUE TITLE
npm install canvas fails: Macro definition of snprintf conflicts with Standard Library function declaration

### ISSUE THREAD

**Issue by neojhz** [Automattic/node-canvas#605]
Hi,

I've installed VS2015 and I have a suspicion that this is due to this.

npm install canvas fails to build with errors:

```text
:\Program Files (x86)\Windows Kits\10\Include\10.0.10150.0\ucrt\stdio.h(1925):
 warning C4005: 'snprintf': macro redefinition (compiling source file ..\src\Ca
nvasPattern.cc) [\node_modules\canvas
\build\canvas.vcxproj]
  C:\Program Files (x86)\Windows Kits\10\Include\10.0.10150.0\ucrt\stdio.h(1925
  ): note: command-line arguments:  see previous definition of 'snprintf' (comp
  iling source file ..\src\CanvasPattern.cc)
C:\Program Files (x86)\Windows Kits\10\Include\10.0.10150.0\ucrt\stdio.h(1927):
 fatal error C1189: #error:  Macro definition of snprintf conflicts with Standa
rd Library function declaration (compiling source file ..\src\CanvasPattern.cc)
 [\node_modules\canvas\build\canvas.v
cxproj]
C:\Program Files (x86)\Windows Kits\10\Include\10.0.10150.0\ucrt\stdio.h(1925):
 warning C4005: 'snprintf': macro redefinition (compiling source file ..\src\Ca
nvasGradient.cc) [...\node_modules\canva
s\build\canvas.vcxproj]
  C:\Program Files (x86)\Windows Kits\10\Include\10.0.10150.0\ucrt\stdio.h(1925
  ): note: command-line arguments:  see previous definition of 'snprintf' (comp
  iling source file ..\src\CanvasGradient.cc)
C:\Program Files (x86)\Windows Kits\10\Include\10.0.10150.0\ucrt\stdio.h(1927):
 fatal error C1189: #error:  Macro definition of snprintf conflicts with Standa
rd Library function declaration (compiling source file ..\src\CanvasGradient.cc
) [...\node_modules\canvas\build\canvas.
vcxproj]
C:\Program Files (x86)\Windows Kits\10\Include\10.0.10150.0\ucrt\stdio.h(1925):
 warning C4005: 'snprintf': macro redefinition (compiling source file ..\src\Ca
nvas.cc) [...\\node_modules\canvas\build\
canvas.vcxproj]
  C:\Program Files (x86)\Windows Kits\10\Include\10.0.10150.0\ucrt\stdio.h(1925
  ): note: command-line arguments:  see previous definition of 'snprintf' (comp
  iling source file ..\src\Canvas.cc)
C:\Program Files (x86)\Windows Kits\10\Include\10.0.10150.0\ucrt\stdio.h(1927):
 fatal error C1189: #error:  Macro definition of snprintf conflicts with Standa
rd Library function declaration (compiling source file ..\src\Canvas.cc) [...\\node_modules\canvas\build\canvas.vcxproj]
  color.cc
  Image.cc
  ImageData.cc
C:\Program Files (x86)\Windows Kits\10\Include\10.0.10150.0\ucrt\stdio.h(1925):
 warning C4005: 'snprintf': macro redefinition (compiling source file ..\src\co
lor.cc) [...\node_modules\canvas\build\c
anvas.vcxproj]
  C:\Program Files (x86)\Windows Kits\10\Include\10.0.10150.0\ucrt\stdio.h(1925
  ): note: command-line arguments:  see previous definition of 'snprintf' (comp
  iling source file ..\src\color.cc)
C:\Program Files (x86)\Windows Kits\10\Include\10.0.10150.0\ucrt\stdio.h(1927):
 fatal error C1189: #error:  Macro definition of snprintf conflicts with Standa
rd Library function declaration (compiling source file ..\src\color.cc) [...\node_modules\canvas\build\canvas.vcxproj]
C:\Program Files (x86)\Windows Kits\10\Include\10.0.10150.0\ucrt\stdio.h(1925):
 warning C4005: 'snprintf': macro redefinition (compiling source file ..\src\Im
age.cc) [...\node_modules\canvas\build\c
anvas.vcxproj]
  C:\Program Files (x86)\Windows Kits\10\Include\10.0.10150.0\ucrt\stdio.h(1925
  ): note: command-line arguments:  see previous definition of 'snprintf' (comp
  iling source file ..\src\Image.cc)
C:\Program Files (x86)\Windows Kits\10\Include\10.0.10150.0\ucrt\stdio.h(1927):
 fatal error C1189: #error:  Macro definition of snprintf conflicts with Standa
rd Library function declaration (compiling source file ..\src\Image.cc) [...\dominantcolor\node_modules\canvas\build\canvas.vcxproj]
  init.cc
```

I've seen this post

http://stackoverflow.com/questions/27754492/vs-2015-compiling-cocos2d-x-3-3-error-fatal-error-c1189-error-macro-definiti

but this requires source changes.

Or is there another way to have this successfully installed?

Thanks.

**Comment by LinusU** [comment]
I don't think that it's possible to come around this without source changes, although it doesn't seem to hard to fix.

This is the offending line [binding.gyp L66](https://github.com/Automattic/node-canvas/blob/60b1fbceb699f8f88d64dd97ab906756abeb8ed3/binding.gyp#L66)

`snprintf` is used in four places:

- [init.cc L56](https://github.com/Automattic/node-canvas/blob/a7c9c0ddb556d8a5c5f683e9f41075ad9cda94e8/src/init.cc#L56)
- [init.cc L58](https://github.com/Automattic/node-canvas/blob/a7c9c0ddb556d8a5c5f683e9f41075ad9cda94e8/src/init.cc#L58)
- [color.cc L455](https://github.com/Automattic/node-canvas/blob/a7c9c0ddb556d8a5c5f683e9f41075ad9cda94e8/src/color.cc#L455)
- [color.cc L460](https://github.com/Automattic/node-canvas/blob/a7c9c0ddb556d8a5c5f683e9f41075ad9cda94e8/src/color.cc#L460)

I don't know the best way to solve this, if possible maybe it could be solved in the `bindings.gyp` file. Otherwise something like this in `init.cc` and `color.cc` should probably solve it.

```c++
// Compatibility with Visual Studio versions prior to 2014
#ifndef snprintf
#define snprintf _snprintf
#endif
```

**Comment by neojhz** [comment]
great. I hope this gets fixed at the source.

**Comment by zorobabel** [comment]
Hey guys, I've posted [a solution](https://github.com/Automattic/node-canvas/issues/619) for this.

**Comment by Zorgatone** [comment]
I have the same issue, fix this please

**Comment by LinusU** [comment]
Pull request very welcome, feel free to try out the solution at the bottom of my last post, and post it as a pull request.

**Comment by LinusU** [comment]
Fixed by #670
